### PR TITLE
Tests to Check Pytorch Hook Sanity With Horovod Training and Zero Code Change

### PR DESCRIPTION
**Pull request by NihalHarish**
### Description of changes:
- Pytorch Horovod script that runs training 
- Tests that check the sanity of the metrics collected by smdebug
- Refactored existing TF Horovod tests to re-use code

**Closed:  See** #216 

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/pytorch/sagemaker_official_container/horovod_mnist.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
+# Future
```
Comment (Vikas-kum) [inline examples/pytorch/sagemaker_official_container/horovod_mnist.py]: sagemaker_official_container, doesn't sound like good name. zero_code_change was earlier name i guess.

File: examples/pytorch/sagemaker_official_container/horovod_mnist.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    # Horovod: initialize library.
+    hvd.init()
+    torch.manual_seed(seed)
+
+    if args.cuda:
+        # Horovod: pin GPU to local rank.
+        torch.cuda.set_device(hvd.local_rank())
+        torch.cuda.manual_seed(seed)
+
+    kwargs = {"num_workers": 1, "pin_memory": True} if args.cuda else {}
```
Comment (Vikas-kum) [inline examples/pytorch/sagemaker_official_container/horovod_mnist.py]: num_workers:1 ? why?

File: examples/tensorflow/sagemaker_official_container/horovod_estimator_mnist.py [hunk 3]
```diff
@@ -35,7 +35,7 @@
 import tensorflow as tf
 from tensorflow import keras
 
-tf.logging.set_verbosity(tf.logging.INFO)
+tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.INFO)
```
Comment (Vikas-kum) [inline examples/tensorflow/sagemaker_official_container/horovod_estimator_mnist.py]: why this change?

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+from smdebug.trials import create_trial
+
+"""
+Tested on current DLAMI p3.8xlarge when run from the main directory
+"""
+
+HOROVOD_MNIST_SCRIPT_NAME = "horovod_mnist.py"
+
+
+def basic_test(out_dir, mode):
```
Comment (Vikas-kum) [inline tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: should basic test also check if losses are saved?.

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+def mode_allworkers_saveall(out_dir, mode):
+    path = build_json(
+        out_dir, include_workers="all", save_all=True, include_collections=["weights", "gradients"]
+    )
+    num_workers = device_count()
+    mode_args = []
+    if mode == "cpu":
+        mode_args += ["--use_only_cpu", "true"]
+    launch_horovod_job(
+        script_file_path=f"examples/pytorch/sagemaker_official_container/{HOROVOD_MNIST_SCRIPT_NAME}",
```
Comment (Vikas-kum) [inline tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: why is this not constant in constant file

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+from torch.cuda import device_count
+
+# First Party
+from smdebug.trials import create_trial
+
+"""
+Tested on current DLAMI p3.8xlarge when run from the main directory
+"""
+
+HOROVOD_MNIST_SCRIPT_NAME = "horovod_mnist.py"
```
Comment (Vikas-kum) [inline tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: need to go in constant file

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    launch_horovod_job(
+        script_file_path=f"examples/pytorch/sagemaker_official_container/{HOROVOD_MNIST_SCRIPT_NAME}",
+        script_args=mode_args,
+        num_workers=num_workers,
+        config_file_path=path,
+        mode=mode,
+    )
+
+    tr = create_trial(out_dir)
+    print(tr.tensor_names())
```
Comment (Vikas-kum) [inline tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: some meaningful print with messages

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    assert len(tr.workers()) == 1
```
Comment (Vikas-kum) [inline tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: comment saying we expect only one worker to be saved because of configuration.

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    assert len(tr.tensor_names()) == 13
+    assert len(tr.tensor(tr.tensor_names(collection="weights")[0]).workers(0)) == 1
+
+
+def test_cpu(out_dir):
+    basic_test(out_dir, "cpu")
+
+
+def test_gpu(out_dir):
+    basic_test(out_dir, "gpu")
```
Comment (Vikas-kum) [inline tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: is this going to run on CI?

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+def test_cpu_allworkers(out_dir):
+    mode_allworkers(out_dir, "cpu")
+
+
+def test_gpu_allworkers(out_dir):
+    mode_allworkers(out_dir, "gpu")
+
+
+def mode_allworkers_saveall(out_dir, mode):
```
Comment (Vikas-kum) [inline tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: this is so similar to mode_allworkers

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    path = build_json(
+        out_dir, include_workers="all", save_all=True, include_collections=["weights", "gradients"]
```
Comment (Vikas-kum) [inline tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: save_all and include_Collections why both?

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py (hunk 4 above)
Comment (Vikas-kum) [reply tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: rename to be something meaningful

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        mode=mode,
+    )
+    tr = create_trial(out_dir)
+    assert len(tr.workers()) == num_workers
+    assert len(tr.tensor_names()) > 25
+    assert len(tr.tensor(tr.tensor_names(collection="weights")[0]).workers(0)) == num_workers
+    assert len(tr.tensor(tr.tensor_names(collection="losses")[0]).workers(0)) == num_workers
+
+
+def test_gpu_allworkers_saveall(out_dir):
```
Comment (Vikas-kum) [inline tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: will it run in CI?

File: examples/tensorflow/sagemaker_official_container/horovod_estimator_mnist.py (hunk 3 above)
Comment (NihalHarish) [reply examples/tensorflow/sagemaker_official_container/horovod_estimator_mnist.py]: `tf.logging` has been moved to `tf.compat.v1.logging` as of TF 1.14 [link](https://stackoverflow.com/questions/38073432/how-to-suppress-verbose-tensorflow-logging) .

But this should ideally be done outside of this PR.
I can revert this.

File: examples/pytorch/sagemaker_official_container/horovod_mnist.py (hunk 1 above)
Comment (NihalHarish) [reply examples/pytorch/sagemaker_official_container/horovod_mnist.py]: Copied over the name from the Tensorflow directory.
I can change it `zero_code_change` and refactor the TF directory structure in a separate PR.

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py (hunk 6 above)
Comment (NihalHarish) [reply tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: Done.

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py (hunk 5 above)
Comment (NihalHarish) [reply tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: Moved to `tests/zero_code_change/horovod_tests/constants.py`

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py (hunk 12 above)
Comment (NihalHarish) [reply tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: Tests run as part of the DLC release pipeline, along with the horovod tests written for tensorflow

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py (hunk 11 above)
Comment (NihalHarish) [reply tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: fixed.

File: tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py (hunk 9 above)
Comment (NihalHarish) [reply tests/zero_code_change/horovod_tests/pytorch/test_distributed_training.py]: Tests run as part of the DLC release pipeline, along with the horovod tests written for tensorflow

File: examples/pytorch/zero_code_change/horovod_mnist.py [hunk 13]
```diff
@@ -0,0 +1,198 @@
+# Future
+from __future__ import print_function
+
```
Comment (vandanavk) [inline examples/pytorch/zero_code_change/horovod_mnist.py]: this test should probably goto amazon-sagemaker-examples too
